safekeeper/src/timeline_eviction.rs

```diff
@@ -43,13 +44,23 @@ impl Manager {
             && self.access_service.is_empty()
             && !self.tli_broker_active.get()
             && !wal_backup_partial::needs_uploading(state, &self.partial_backup_uploaded)
-            && self
-                .partial_backup_uploaded
-                .as_ref()
-                .unwrap()
-                .flush_lsn
-                .segment_number(self.wal_seg_size)
-                == self.last_removed_segno + 1
+            // all (full) segments before the uploaded partial must be committed
+            // and uploaded.
+            && state.cfile_backup_lsn.segment_number(self.wal_seg_size)
+                == state.flush_lsn.segment_number(self.wal_seg_size);
+        if !ready {
+            return false;
+        }
+        // and all local segments which can be deleted are garbage
+        // collected, there must be only one partial segment on disk left
+        let n_segs_on_disk = match count_segments(self.tli.timeline_dir()).await {
+            Ok(n) => n,
+            Err(e) => {
+                warn!("failed to count segments: {:?}", e);
+                return false;
+            }
+        };
+        ready && n_segs_on_disk == 1
```

Comment:
> Also slightly tweak eviction condition: all full segments before
flush_lsn must be uploaded (and committed) and there must be only one
segment left on disk (partial). This allows to evict timelines which
started not on the first segment and didn't fill the whole
segment (previous condition wasn't good because last_removed_segno was
0).

FYI this is not needed because `last_removed_segno` is always updated on the first iteration.

More specifically, `update_wal_removal` spawns a task which calculates a horizon and deletes all segments before it. Afterwards it updates `last_removed_segno` to match the horizon.

I'd rollback this change because calling `count_segments()` in `ready_for_eviction()` is not very efficient.

Reply:
Cool, wasn't aware of that. Rolled back. I made another small change: `self.partial_backup_uploaded.is_some()` check is redundant because `!wal_backup_partial::needs_uploading(state, &self.partial_backup_uploaded)` does it, please have a look.